### PR TITLE
Stat setup

### PR DESCRIPTION
**Pull request by pecasdev**
Added new Stat concept
- Stats represent row data that the frontend will convert into a visual, probably a graph.
- Stat computations are cached and re-fetched on request only if a "staleness" period has passed.
- Each unique stat has its own implementation of manipulating answer data into relevant stat row data. The current flow is "fetch every answer set in the database" -> "do something with those answers to fill out row data" -> "cache row data for at least an hour before refetching". This felt like an easier solution to scale than defining custom SQL queries for each stat and performing data manipulation in the database layer. 
- Stats can be queried via a new endpoint `/stat?id={stat.id}`. Currently only "job-title" exists in the known stats, I'll add the rest after we discuss what stats we want the final product to have.

### CODE REVIEW

File: backend/src/main/java/com/jobwait/spring/SpringEndpoints.java [hunk 1]
```diff
@@ -18,13 +18,17 @@
...
 		try {
-			responseNode.set("data", func.get());
-			return new ResponseEntity<ObjectNode>(responseNode, HttpStatus.OK);
+			ObjectNode data = func.get();
+			if (data == null) {
+				return new ResponseEntity<ObjectNode>(HttpStatus.OK);
+			} else {
+				responseNode.set("data", func.get());
+				return new ResponseEntity<ObjectNode>(responseNode, HttpStatus.OK);
+			}
```
Comment (pecasdev) [inline backend/src/main/java/com/jobwait/spring/SpringEndpoints.java:31]: relevant when submitting answers, getting back `{data: null}` is redundant

File: backend/src/main/java/com/jobwait/spring/SpringEndpoints.java [hunk 2]
```diff
@@ -55,4 +59,18 @@ public static ResponseEntity<ObjectNode> submitUserAnswers(@RequestParam("at") S
...
+	}
+	/*
+	 * 
+	 * frontend can query a specific stat by id
+	 * backend will send a dictionary of string -> list[int] with rowdata for that
+	 * stat
+	 * frontend takes the rows and a provided schema and renders it using graph
+	 * library
+	 * 
+	 */
```
Comment (pecasdev) [inline backend/src/main/java/com/jobwait/spring/SpringEndpoints.java]: remove this

File: backend/src/main/java/com/jobwait/stat/Stats.java [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+import org.springframework.http.HttpStatus;
+
+import com.jobwait.domain.Stat;
+import com.jobwait.fault.FaultException;
+import com.jobwait.stat.impl.JobTitleStat;
+
+public class Stats {
+    // touch this stuff
+    public static List<Stat> known = List.of(
+            new JobTitleStat("job-title"));
```
Comment (pecasdev) [inline backend/src/main/java/com/jobwait/stat/Stats.java:14]: "job-title" should be defined inside JobTitle, it feels redundant to pass it as an argument to the class of the same name. I did this on purpose because if you want to know at a glance what id is used for a certain stat, you can go to Stats and see them all here. If I put it in the class, you have to go into each class to see their ids. I think the convenience is worth the redundancy, but let me know your thoughts.

File: backend/src/main/java/com/jobwait/control/RequestController.java [hunk 4]
```diff
@@ -1,14 +1,17 @@
 package com.jobwait.control;
 
 import java.util.List;
+import java.util.Map;
```
Comment (danielperev) [inline backend/src/main/java/com/jobwait/control/RequestController.java]: whats the use of this import?

File: backend/src/main/java/com/jobwait/persistence/PostgresController.java [hunk 5]
```diff
@@ -153,6 +153,25 @@ public User createUserFromAuthId(String authHash) {
...
+    public List<List<Answer>> getAllAnswerSets() {
+        try {
+            Connection connection = getConnection();
+
+            String answerColumnValues = String.join(", ", Questions.knownQuestionKeys);
+            String statementText = "SELECT %s FROM answers".formatted(answerColumnValues);
+            PreparedStatement statement = connection.prepareStatement(statementText);
+
+            ResultSet resultSet = statement.executeQuery();
+            List<List<Answer>> answerSets = PersistenceUtil.resultSetRowsToAdaptedRows(resultSet,
```
Comment (danielperev) [inline backend/src/main/java/com/jobwait/persistence/PostgresController.java:166]: i really hate this `List<List<Answer>>`, i understand why its there but i really dont like reading it. anyway at all to make use of Java's new Optional?

File: backend/src/main/java/com/jobwait/stat/StatRefreshOfficer.java [hunk 6]
```diff
@@ -0,0 +1,33 @@
...
+
+import java.time.Duration;
+import java.time.Instant;
+import java.util.Map;
+import java.util.function.Function;
+import java.util.stream.Collectors;
+
+import com.jobwait.domain.Stat;
+
+public class StatRefreshOfficer {
```
Comment (danielperev) [inline backend/src/main/java/com/jobwait/stat/StatRefreshOfficer.java]: Officer is a new one, I would be interested to hear the context behind the choice

File: backend/src/main/java/com/jobwait/stat/impl/JobTitleStat.java [hunk 7]
```diff
@@ -0,0 +1,41 @@
...
+        List<List<Answer>> answers = this.fetchAllAnswerSets();
+        for (List<Answer> answerSet : answers) {
+            for (Answer answer : answerSet) {
+                if (answer.questionKey.equals("jobtitle")) {
+                    String title = answer.answerValueAsString();
+                    if (title != null) {
+                        this.addOrIncrementRow(answer.answerValueAsString());
+                    }
+                }
+            }
```
Comment (danielperev) [inline backend/src/main/java/com/jobwait/stat/impl/JobTitleStat.java:38]: if my previous question about Optional makes no sense (because I misunderstood why we have List of List) disregard this: if we had optional here we wouldnt have to do the double for loop

File: backend/src/main/java/com/jobwait/stat/Stats.java (hunk 3 above)
Comment (danielperev) [reply backend/src/main/java/com/jobwait/stat/Stats.java:14]: I think that's fair. Moreover, I think it would be useful if potentially (I don't know why) we had the same stat but with a different id. Either way I think its ok for now at least.

File: backend/src/main/java/com/jobwait/stat/Stats.java [hunk 8]
```diff
@@ -0,0 +1,35 @@
...
+
+    // don't touch this stuff
+    private static StatRefreshOfficer statRefreshOfficer = new StatRefreshOfficer();
+
+    public static Stat statFromId(String statId) {
+        List<Stat> matches = known.stream().filter(s -> s.id.equals(statId)).toList();
+        if (matches.size() == 0) {
+            throw new UnknownStatId(statId);
+        }
+        Stat stat = matches.getFirst();
```
Comment (danielperev) [inline backend/src/main/java/com/jobwait/stat/Stats.java]: is there no `getOrElse` or `getFirstOrElse` ?

File: backend/src/main/java/com/jobwait/stat/StatRefreshOfficer.java [hunk 9]
```diff
@@ -0,0 +1,33 @@
...
+    // constants
+    private static Duration statLifetime = Duration.ofMinutes(60);
+
+    // stat id to timestamp
+    private static Map<String, Instant> generateIdStampMap() {
+        return Stats.known.stream().map(s -> s.id)
+                .collect(Collectors.toMap(Function.identity(), _ -> Instant.now(), (_, _) -> {
```
Comment (danielperev) [inline backend/src/main/java/com/jobwait/stat/StatRefreshOfficer.java]: do you prefer `Function.identity()` over `s -> s` ? just a question of preference I suppose

File: backend/src/main/java/com/jobwait/stat/StatRefreshOfficer.java [hunk 10]
```diff
@@ -0,0 +1,33 @@
...
+                    throw new RuntimeException("impossible error");
+                }));
+    }
+
+    private Map<String, Instant> statIdToLastRefreshStamp = generateIdStampMap();
+
+    public void refreshStatIfNeeded(String statId) {
```
Comment (danielperev) [inline backend/src/main/java/com/jobwait/stat/StatRefreshOfficer.java]: the `IfNeeded` here is a bit ambiguous, what about `refreshOldStat` ? idk just some way to indicate that its only refreshing old stat

**Review by danielperev — CHANGES_REQUESTED** [review]
some minor change requests / clarifications but i think its good!

File: backend/src/main/java/com/jobwait/control/RequestController.java (hunk 4 above)
Comment (pecasdev) [reply backend/src/main/java/com/jobwait/control/RequestController.java]: vscode must have missed this one, I'll remove it

File: backend/src/main/java/com/jobwait/persistence/PostgresController.java (hunk 5 above)
Comment (pecasdev) [reply backend/src/main/java/com/jobwait/persistence/PostgresController.java:166]: I don't think Optional fits here. The inner list represents the answers belonging to a single user, so this is a list of answer sets. I think the type makes sense here. Sure I could flatten it into a List<Answer> but maybe some stats care about the correlation between answers given by the same user (i.e. "how many full-time employees are remote?") so I think it's useful to leave it as List<List<Answer> for more information.

I agree, at first I also hated this type because nested generic classes smell funny, but I think it's grown on me. The only alternative I see here would be creating an Answers wrapper class that represents a set of Answers but I literally just removed that and I think it gets in the way of Questions being the SSOT (single source of truth) if we also have to maintain a separate Answers class too.

In conclusion, List<List<Answer>> smells funky but when the shape of the data is considered, I think it's acceptable for the time being. I hope this explanation makes you hate the type signature less.

File: backend/src/main/java/com/jobwait/stat/StatRefreshOfficer.java (hunk 6 above)
Comment (pecasdev) [reply backend/src/main/java/com/jobwait/stat/StatRefreshOfficer.java]: I got bored of Manager and Controller, simply wanted to spice things up so I'm inspired when I see the code. Maybe I should have picked Controller for consistency's sake (heavily considering that now). Ok yeah I'll change it to Controller

File: backend/src/main/java/com/jobwait/stat/impl/JobTitleStat.java (hunk 7 above)
Comment (pecasdev) [reply backend/src/main/java/com/jobwait/stat/impl/JobTitleStat.java:38]: I think you might have misunderstood the shape of the data. I responded to your other comment. I'll mark this as resolved (or you can too) if the question no longer applies.

File: backend/src/main/java/com/jobwait/persistence/PostgresController.java (hunk 5 above)
Comment (danielperev) [reply backend/src/main/java/com/jobwait/persistence/PostgresController.java:166]: Ok so this is a List of answers per user, but the Answer type does not store any reference to a user so how is this distinguished at all? besides the fact that each list of answers is distinct which I guess is a somewhat useful heuristic.

I am not sure how making a wrapper class gets in the way of SSOT, SSOT does not have to mean one class stores everything, in our system answers are responses to questions, they are both within the realm of the truth, not just the questions being asked.

File: backend/src/main/java/com/jobwait/stat/Stats.java (hunk 8 above)
Comment (pecasdev) [reply backend/src/main/java/com/jobwait/stat/Stats.java]: rare java W
91a8538460ba624b1112c988479ef46d61dafa53

File: backend/src/main/java/com/jobwait/stat/StatRefreshOfficer.java (hunk 9 above)
Comment (pecasdev) [reply backend/src/main/java/com/jobwait/stat/StatRefreshOfficer.java]: Funny that the "more readable" version is like triple the character count. I have no strong opinion on this, I'd be fine either way

File: backend/src/main/java/com/jobwait/stat/StatRefreshOfficer.java (hunk 10 above)
Comment (pecasdev) [reply backend/src/main/java/com/jobwait/stat/StatRefreshOfficer.java]: I wanted to avoid that because the function is not always refreshing, it's only refreshing if it's needed. Maybe `refreshStatIfStale` is more specific? I'll change it to that, let me know if you prefer something else

File: backend/src/main/java/com/jobwait/control/RequestController.java (hunk 4 above)
Comment (pecasdev) [reply backend/src/main/java/com/jobwait/control/RequestController.java]: 64059a8c4367666cd481ffbe59d0dd042c4b21bd

File: backend/src/main/java/com/jobwait/persistence/PostgresController.java (hunk 5 above)
Comment (pecasdev) [reply backend/src/main/java/com/jobwait/persistence/PostgresController.java:166]: > but the Answer type does not store any reference to a user so how is this distinguished at all?

Good point, but right now the concept of a User doesn't mean much besides just authentication. I don't think it will ever mean more than that. After all, any important answer we want to consider for the sake of a Stat will be implemented as another Question, not be a property on the User. So in that regard, I don't think it matters *who* answered the questions so long as we know which answers belong to the same User

> I am not sure how making a wrapper class gets in the way of SSOT

If we have an Answers class with fields for each question, we have to maintain that when we add a new question. It's no longer "update the Questions class" it's now, "update the Questions class and update the Answers class to match". Sure you might say "well Peter, it's just another line change, so what" but I say it's no longer SSOT.

Unless the Answers wrapper class doesn't have fields per question and just has a single property which is a List<Answer> which contains every answer a user made, but idk that feels redundant to me.

File: backend/src/main/java/com/jobwait/stat/StatRefreshOfficer.java (hunk 6 above)
Comment (pecasdev) [reply backend/src/main/java/com/jobwait/stat/StatRefreshOfficer.java]: f1a9bf82f7909cf16187bcb4239aa49ea4a02f92

File: backend/src/main/java/com/jobwait/persistence/PostgresController.java (hunk 5 above)
Comment (danielperev) [reply backend/src/main/java/com/jobwait/persistence/PostgresController.java:166]: > So in that regard, I don't think it matters who answered the questions so long as we know which answers belong to the same User

I agree, what I mean is that a List of List is easy to accidentally manipulate, whereas a Map of User id to List of answer would be more difficult, but this is pedantic. We are on the same page lets just leave it for now.

> Unless the Answers wrapper class doesn't have fields per question and just has a single property which is a List which contains every answer a user made, but idk that feels redundant to me.

Perhaps, I guess its not a big deal although in the end `List<Answers>` would be marginally clearer than `List<List<Answer>>`

File: backend/src/main/java/com/jobwait/stat/StatRefreshOfficer.java (hunk 9 above)
Comment (danielperev) [reply backend/src/main/java/com/jobwait/stat/StatRefreshOfficer.java]: just leave it then honestly, no biggie

**Review by danielperev — APPROVED** [review]
🚀 😲

File: backend/src/main/java/com/jobwait/persistence/PostgresController.java (hunk 5 above)
Comment (pecasdev) [reply backend/src/main/java/com/jobwait/persistence/PostgresController.java:166]: alright you convinced me, List<Answers> is more readable (at the risk of adding class bloat?), I'll make the change

Comment (pecasdev) [reply backend/src/main/java/com/jobwait/persistence/PostgresController.java:166]: nvm it's a pain in the ass cause PostgresAdapter expects to return a List<A> for a type A, I can't just make it return a new wrapper class.
